update for generating summary table

ZechangSun opened the pull request:
Dear X,
       This PR is about generating summary table for a given list of `dla_cnn.data_model.Sigthline.Sightline` objects. The detail of this part can be seen at `dla_cnn\desi\preprocess.py\generate_summary_table`. I changed something of `dla_cnn\data_model\Sigthline.py` and `dla_cnn\desi\DesiMock.py` to generate the summary table. Besides I have uploaded a Jupyter notebook `dla_cnn\desi\nb\generate summary table.ipynb` to show the summary table we got.

Best,
Zechang Sun

ZechangSun committed:
update for generating summary table

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Update Sightline.py

ZechangSun committed:
Add files via upload

ZechangSun committed:
Update preprocess.py

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -219,3 +219,41 @@ def estimate_s2n(sightline):
+    None
+    
+    """
+    import csv
```
put this (and all imports) at the top of the module

ZechangSun replied:
OK , I have updated that. Thanks!

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -219,3 +219,41 @@ def estimate_s2n(sightline):
+
+def generate_summary_table(sightlines, output_dir, mode = "w"):
+    """
+    Generate a csv file to store some necessary information of the given sightlines. The necessary information means the id, z_qso,
```
Excellent doc string!

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -219,3 +219,41 @@ def estimate_s2n(sightline):
+    Generate a csv file to store some necessary information of the given sightlines. The necessary information means the id, z_qso,
+    s/n of thelymann forest part(avoid dlas+- 3000km/s), the wavelength range and corresponding pixel number of each channel. And the csv file's format is like:
+    
+    id(int)， z_qso(float), s2n(float), wavelength_start_b(float), wavelength_end_b(float), pixel_start_b(int), pixel_end_b(int), wavelength_start_r(float), wavelength_end_r(float), pixel_start_r(int), pixel_end_r(int), wavelength_start_z(float), wavelength_end_z(float), pixel_start_z(int), pixel_end_z(int)
```
Make sure this includes everything that @zoujiaqi99 needs for generating her training sets.

profxj replied:
The obvious thing missing is information about the DLAs.  
Is that held in a separate Table that @zoujiaqi99 can access?
Or do we need to include such information here?  You two should discuss.

ZechangSun replied:
I have discussed with @zoujiaqi99 , and we chose to the DLAs' information to the summary table, detail of this can be seen in doc string of `generate_summary_table`, thanks for your advice! By the way ,  a new jupyter notebook has been uploaded to `desi/nb/Generate Summary Table.ipynb`, you can check the summary table here. Hope for your advice, thanks!

profxj reviewed (changes requested):
Nice work!  My main request is that you copy the code you have in the Notebook
into a module so that we can remake the Summary table (for any set of mock 
spectra) without using a Notebook.

Notebooks are good for testing but not for running repeatedly.

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Delete generate summary table.ipynb

ZechangSun committed:
Add files via upload

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -255,5 +258,67 @@ def generate_summary_table(sightlines, output_dir, mode = "w"):
+        """
+        #read data from fits file
+        file_path = join(path,str(num))
+        spectra = join(file_path,"spectra-16-%s.fits"%num)
```
why is the `16` hard-coded?  
More important, what does it mean?

ZechangSun replied:
That is due to the dataset we use, we use the mock spectrum v9.16, so the fits files' name is in this format, e.g. "spectra-16-700.fits". Maybe we need to set `16` as a parameter.

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -225,10 +228,11 @@ def generate_summary_table(sightlines, output_dir, mode = "w"):
     "wavelength_start_b" means the start wavelength value of b channel, "wavelength_end_b" means the end wavelength value of b channel, "pixel_start_b" means the start pixel number of b channel, "pixel_end_b" means the end pixel number of b channel
-    so do the other two channels.
+    so do the other two channels.Besides, "dlas_col_density" means the col_density array of the sightline, and "dlas_central_wavelength" means the central wavelength array means the central wavelength array of the given sightline. Due to the uncertainty of the dlas' number, we chose to use str format to store the two arrays,
+    each array is written in the format like "value1,value2, value3", and one can use `str.split(",")` to get the data, the column density and central wavelength which have the same index in the two arrays corrspond to the same dla.
```
I can understand why you did this (and probably had to), but
it will make using the Table for DLAs pretty difficult, if not impossible.

Maybe we just need a separate DLA table.
Anyhow, I'm ok with this for now and we can modify it later as need be.

ZechangSun replied:
Sure.

ZechangSun replied:
I will do it later.

profxj reviewed (approved):
Address the comment about `16` and I'll approve.

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Update Generate Summary Table.ipynb

profxj reviewed (changes requested):
Move the code in the `Generate Summary Table` into a module,
ideally one that can be run from a script (see `dla_cnn/scripts/analyze_sl.py`
and I'll approve.

Thanks!

ZechangSun committed:
Create generate_summary_table.py

ZechangSun committed:
Create __init__.py

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Update generate_summary_table.py

ZechangSun commented:
> 
> 
> Move the code in the `Generate Summary Table` into a module,
> ideally one that can be run from a script (see `dla_cnn/scripts/analyze_sl.py`
> and I'll approve.
> 
> Thanks!

Hi, Prof. X. I have just moved the `Generate Summary Table` into the `desi\scripts\generate_summary_table.py` module, and it can be directly run from a script. Sorry, I have also added more criteria in the code to identify whether a spectrum can be used for sample after discussing with @zoujiaqi99 .  Sorry that I didn't notice this review in time. I will check my gmail regularly next time.

profxj reviewed (approved):
Great!
